pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -594,27 +598,106 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
+//     use the original port number.
+//
+//   - (DNS) clients use ephemeral source ports, i.e., the port can be different in every
+//     request. Typically, a DNS resolver library uses the same ephemeral port only for requests
+//     from a single "gethostbyname" API call, or equivalent.
+//
+//   - To be able to receive responses to the ephemeral source port, we must have a socket bound to
+//     that address:port (for UDP), or a connection from that address:port to the DNS server
+//     address:port (for TCP).
+//
+//   - This leads to a new DNS client and socket for every different original source address -
+//     ephemeral port pair we see. We also need to make sure these were actually used to communicate
+//     with the DNS server, so we use the whole 5-tuple as a key.
+//
+// Why can't we keep DNS clients pooled and ready to receive traffic between client requests?
+//
+//   - We have no guarantees that the source pod will keep on using the same ephemeral port in
+//     future. We've had upstream socket bind errors (in Envoy, where we have operated in this mode
+//     for years already) when a client pod has rapidly cycled through its ephemeral port space,
+//     e.g. when performing netperf CRR or similar performance tests.
+//
+//   - We could try to keep the client and its bound socket around for some minimal time to save
+//     resources when a DNS resolver is enumerating through its domain suffix list, where it seems
+//     likely that the same source ephemeral port is going to be reused until the resolver gets an
+//     actual result with an IPv4/6 address or quits trying. It might be safe to close the client
+//     socket only after a response with the `A`/`AAAA` records have been passed back to the pod,
+//     or after a timeout of a few milliseconds. This would be something we currently don't do and
+//     is prone to socket bind errors, so this is left for a later exercise.
+//
+//   - So the client socket can not be left lingering around, as it causes network traffic destined
+//     for the source pod to be intercepted to the dnsproxy, which is exactly what we want but only
+//     until a DNS response has been received.
+func setSoMarks(fd int, ipv4 bool, secId identity.NumericIdentity) error {
+	// Set SO_MARK to allow datapath to know these upstream packets from an egress proxy
+	mark := linux_defaults.MagicMarkEgress
 	mark |= int(uint32(secId&0xFFFF)<<16 | uint32((secId&0xFF0000)>>16))
 	err := unix.SetsockoptInt(fd, unix.SOL_SOCKET, unix.SO_MARK, mark)
 	if err != nil {
 		return fmt.Errorf("error setting SO_MARK: %w", err)
 	}
+
+	// Rest of the options are only set in the transparent mode.
+	if !option.Config.DNSProxyEnableTransparentMode {
+		return nil
+	}
+
+	// Set IP_TRANSPARENT to be able to use a non-host address as the source address
+	if err := transparentSetsockopt(fd, ipv4, !ipv4); err != nil {
```

Comment:
Same [comment](https://github.com/cilium/cilium/pull/30216/files#r1452525997) as v1.13 backport: Differently from v1.14 onwards, here we are calling `transparentSetsockopt` and thus also setting `IP_RECVORIGDSTADDR`. I don't think this makes any actual difference, but maybe it is better to just set `IP_TRANSPARENT` or `IPV6_TRANSPARENT` according to the `ipv4` flag.